torch/csrc/jit/codegen/cuda/ops/normalization.cpp

```diff
@@ -7,6 +7,64 @@ namespace jit {
 namespace fuser {
 namespace cuda {
 
+int positiveAxis(int axis, int ndims) {
+  return (axis > 0) ? axis : (ndims + axis);
+}
+
+Val* numFeatures(TensorView* x, const std::vector<int>& dims, int ndims) {
+  Val* num_features = IrBuilder::create<Double>(x->container(), 1);
+  for (const auto dim : dims) {
+    const int axis = positiveAxis(dim, ndims);
+    num_features = mul(num_features, x->domain()->domain()[axis]->extent());
+  }
+  return num_features;
+}
+
+TensorView* mean(TensorView* x, const std::vector<int>& dims, bool keepdim) {
+  TORCH_INTERNAL_ASSERT(x != nullptr, "Input is invalid.");
+
+  const int kNumberOfDims =
+      TensorDomain::noReductions(x->getMaybeRFactorDomain()).size();
+
+  auto sum_x = sum(x, dims, keepdim);
+  auto y = div(sum_x, numFeatures(x, dims, kNumberOfDims));
+  return y;
+}
+
+TensorView* variance(
+    TensorView* x,
+    const std::vector<int>& dims,
+    bool unbiased,
+    bool keepdim) {
+  TORCH_INTERNAL_ASSERT(x != nullptr, "Input is invalid.");
+
+  const int kNumberOfDims =
+      TensorDomain::noReductions(x->getMaybeRFactorDomain()).size();
+
+  auto bcast_mean = mean(x, dims, true /* keepdim */);
+  auto x_mean_sub = sub(x, bcast_mean);
+  auto x_mean_sub_sq = mul(x_mean_sub, x_mean_sub);
```

Comment:
Can you use welford here?

Reply:
Yes. I guess when you need both mean and variance, you would be calculating the mean twice.